File: offer/16.go
/*
 * @Author: lmio
 * @Date: 2023-04-10 08:18:09
 * @LastEditTime: 2023-04-10 08:23:31
 * @FilePath: /leetcode/offer/16.go
 * @Description:剑指 Offer 16. 数值的整数次方
 */
package offer

func MyPow(x float64, n int) float64 {
	res := 1.0
	if n < 0 {
		x = 1.0/x
	}
	for n > 0 {
		if n & 1 == 1 {
			res *= x
		}
		x *= x
		n = n >> 1
	}
	return res
}

File: code/16.py
'''
Author: lmio 2091319361@qq.com
Date: 2023-07-10 12:42:16
LastEditors: lmio 2091319361@qq.com
Description: 16. 最接近的三数之和
'''

import bisect
from typing import List


def threeSumClosest(nums: List[int], target: int) -> int:
    nums.sort()
    n = len(nums)
    gap = float('inf')
    res = 0
    for i in range(n):
        for j in range(i+1, n-1):
            sum = nums[i] + nums[j]
            sum += findClosest(nums[j+1:], target-sum)
            _gap = abs(target - sum)
            if _gap < gap:
                gap = _gap
                res = sum
    return res

def findClosest(nums: List[int], target: int):
    insert_pos = bisect.bisect_left(nums, target)
    
    if insert_pos == 0:
        return nums[0]
    elif insert_pos == len(nums):
        return nums[-1]
    else:
        left_diff = abs(target - nums[insert_pos - 1])
        right_diff = abs(target - nums[insert_pos])
        
        if left_diff <= right_diff:
            return nums[insert_pos - 1]
        else:
            return nums[insert_pos]

print(threeSumClosest([1,1,1,0], -100))